### PR TITLE
Use new JSON Streaming methods for Invoke-RestMethod

### PR DESCRIPTION
**Pull request by strawgate**
# PR Summary

This PR is a "child" PR to: https://github.com/PowerShell/PowerShell/pull/15695

This PR is in good shape but needs tests for the additional parameter on invoke-restmethod for -ashashtable

Essentially if the streaming approach from that PR is approved we can make new methods which take a stream directly instead of the string body. We can then wire up invoke-restmethod so that it takes the response stream and sends it directly to newtonsoft.

This should improve all use-cases of invoke-restmethod but will particularly improve cases where the source API has an array as its root element.

## PR Context

This should greatly reduce the memory usage of the invoke-restmethod cmdlets and somewhat reduce the execution time/cpu time.

Benchmarks for a 100MB JSON Response with Array root element (Time to first byte from the API is about 8s so for realistic times subtract 8000ms from each runtime number)

7.1.3:
- Memory Usage: **1597 MB**
- Runtime: **20679.9797 ms**
- Est Runtime: **12679.9797 ms**

ThisPR:
- Memory Usage: 718.55 MB
- Runtime: 16323.1985 ms
- Est Runtime: 8323.1985 ms

ThisPR -AsHashtable:
- Memory Usage: **415.41 MB**
- Runtime: **13592.6211 ms**
- Est Runtime: **5592.6211 ms**

Benchmarks for a 100MB JSON response with Object root element
7.1.3
- Memory Usage: 1672.84765625 MB
- Runtime: 12771.3602 ms
- Est Runtime: 12771.3602 ms

This PR:
- Memory Usage: 1489.90625 MB
- Runtime: 12222.5628 ms

This PR -AsHashtable
- Memory Usage: 1146.078125 MB
- Runtime: 11367.9809 ms

Benchmarks for a 12MB JSON Response
7.1.3:
- Memory Usage: 268.9921875 MB
- Runtime: 2553.9797 ms

ThisPR:
- Memory Usage: 100.05078125 MB
- Runtime: 1950.1985 ms

Runtime is misleading as I am hitting a real API

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
@strawgate I believe we have had a consensus that we froze current Json cmdlets based on NewtonSoft and implement new ones based on .Net API.

**Comment by strawgate** [comment]
I looked around a fair amount for information on the status of System.Text.Json and all I found was a couple 2 year old PRs that haven't moved.

My goal was to offer a highly-compatible set of changes that make a big impact on performance while we wait for System.Text.Json. The changes outlined here cut memory usage to 1/4 of what it was previously if the root element is an array and ~25% if it's an object.

If you guys don't want to merge the changes that's totally fine!

**Comment by iSazonov** [comment]
@strawgate The main problem is that the MSFT team is too small to review everything PR quickly.
They introduced Working Groups to get more developers from the community. If you want to work on commandlets in this repository, you can join one of the working groups.

The current strategic direction in the C# community is migration from NewtonSoft.JsonNet to System.Text.Json API. NewtonSoft has been found to be out of development. All new developers should be on System.Text.Json API.

It doesn't make sense to invest much in code based on NewtonSoft.JsonNet, only security issues, simple bugs/perfs. 

If you are interested in working on these commandlets, the current conclusion is to implement new commandlets based on the System.Text.Json API as experimental features. You can grab my PRs then I could review your PRs.

**Comment by strawgate** [comment]
I've been using https://www.mockaroo.com/ to generate random json data with numbers, dates, etc and have been using http://www.jsondiff.com/ to do a semantic compare between the 7.1.3 cmdlets and the ones in this PR. So far so good!

**Comment by strawgate** [comment]
> @strawgate The main problem is that the MSFT team is too small to review everything PR quickly.
> They introduced Working Groups to get more developers from the community. If you want to work on commandlets in this repository, you can join one of the working groups.
> 
> The current strategic direction in the C# community is migration from NewtonSoft.JsonNet to System.Text.Json API. NewtonSoft has been found to be out of development. All new developers should be on System.Text.Json API.
> 
> It doesn't make sense to invest much in code based on NewtonSoft.JsonNet, only security issues, simple bugs/perfs.
> 
> If you are interested in working on these commandlets, the current conclusion is to implement new commandlets based on the System.Text.Json API as experimental features. You can grab my PRs then I could review your PRs.

I can't find information on joining a specific working group, is that info available somewhere?

**Comment by iSazonov** [comment]
> I can't find information on joining a specific working group, is that info available somewhere?

You could e-mail directly @joeyaiello and @SteveL-MSFT.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs [hunk 1]
```diff
@@ -210,6 +259,56 @@ public static object ConvertFromJson(string input, bool returnHashtable, int? ma
             }
         }
 
+        private static List<object> DeserializeRootArrayHelper(JsonReader reader, JsonSerializer serializer, bool returnHashtable, out ErrorRecord error)
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:262]: Add description of the method and parameters

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
preliminary review

**Comment by TravisEz13** [comment]
Converted to draft, please marks as ready for review when the parent PR is merged.

**Comment by strawgate** [comment]
I think this PR will be abandoned as there is no interest in improving the performance of the existing cmdlets due to backwards compatibility concerns.

**Comment by iSazonov** [comment]
> I think this PR will be abandoned as there is no interest in improving the performance of the existing cmdlets due to backwards compatibility concerns.

All this could be in new experimental cmdlets based on System.Text.Json API.

**Pull request closed without merging**